app/models/user.rb

```diff
@@ -1,13 +1,73 @@
 class User < ActiveRecord::Base
+  
+  attr_accessor :remember_token, :activation_token, :reset_token 
+
+  before_save :downcase_email
+  before_create :create_activation_digest
+
   has_many :posts, dependent: :destroy
   has_secure_password
 
-  before_save { self.email = email.downcase }
-
   validates :name, presence: true, length: { maximum: 50 }
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   validates :email, presence: true, length: { maximum: 255 },
-    format: { with: VALID_EMAIL_REGEX },
-    uniqueness: { case_sensitive: false }
+            format: { with: VALID_EMAIL_REGEX },
+            uniqueness: { case_sensitive: false }
   validates :password, presence: true, length: { minimum: 6 }
-end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end 
+
+  def User.digest(string)
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+                                                  BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
+  end 
+
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil? 
+    BCrypt::Password.new(digest).is_password?(token)
+  end 
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
+  end 
+
+  def forget 
+    update_attribute(:remember_digest, nil)
+  end 
+
+  def activate 
+    update_attribute(:activated, true)
+    update_attribute(:activate_at, Time.zone.now)
+  end 
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end 
+
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end 
+
+  private 
+
+  def downcase_email
+    self.email = email.downcase
+  end 
+
+  def create_activation_digest
+    self.activation_token = User.new_token
+    self.activation_digest = User.digest(activation_token)
+  end 
+
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest, User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
```

Comment:
would recommend `update_attributes` here as well and combine the arguments of these two into one hash